lib/Dialect/Handshake/HandshakeTypes.cpp

```diff
@@ -115,58 +262,41 @@ static Type parseChannelAfterLess(AsmParser &odsParser) {
                         "which is to be a Type");
     return nullptr;
   }
+
+  // TODO: This check is also called in verify. We should consider refactoring
+  // this.
   if (failed(checkChannelData(emitError, *dataType)))
     return nullptr;
 
-  // Parse variable 'extraSignals'
-  extraSignalsStorage = [&]() -> FailureOr<SmallVector<ExtraSignal::Storage>> {
-    SmallVector<ExtraSignal::Storage> storage;
-
-    if (!odsParser.parseOptionalComma()) {
-      auto parseSignal = [&]() -> ParseResult {
-        auto &signal = storage.emplace_back();
-
-        if (odsParser.parseKeywordOrString(&signal.name) ||
-            odsParser.parseColon() || odsParser.parseType(signal.type))
-          return failure();
-
-        // Attempt to parse the optional upstream symbol
-        if (!odsParser.parseOptionalLParen()) {
-          std::string upstreamSymbol;
-          if (odsParser.parseKeywordOrString(&upstreamSymbol) ||
-              upstreamSymbol != UPSTREAM_SYMBOL || odsParser.parseRParen())
-            return failure();
-          signal.downstream = false;
-        }
-        return success();
-      };
-
-      if (odsParser.parseLSquare() ||
-          odsParser.parseCommaSeparatedList(parseSignal) ||
-          odsParser.parseRSquare())
-        return failure();
-    }
+  SmallVector<ExtraSignal::Storage> extraSignalsStorage;
+  if (!odsParser.parseOptionalComma()) {
+    // Parsed literal ','
+    // The channel has extra bits
 
-    return storage;
-  }();
-  if (failed(extraSignalsStorage)) {
-    odsParser.emitError(odsParser.getCurrentLocation(),
-                        "failed to parse ChannelType parameter 'extraSignals' "
-                        "which is to be a ArrayRef<ExtraSignal>");
-    return nullptr;
+    // Parse '['
+    if (odsParser.parseLSquare())
+      return nullptr;
+
+    if (failed(parseExtraSignals(emitError, odsParser, extraSignalsStorage)))
+      return nullptr;
+
+    // Parse ']'
+    if (odsParser.parseRSquare())
+      return nullptr;
   }
 
   // Parse literal '>'
   if (odsParser.parseGreater())
     return {};
 
+  SmallVector<ExtraSignal> extraSignals;
   // Convert the element type of the extra signal storage list to its
   // non-storage version (these will be uniqued/allocated by ChannelType::get)
-  SmallVector<ExtraSignal> extraSignals;
-  for (const ExtraSignal::Storage &signalStorage : *extraSignalsStorage)
+  for (const ExtraSignal::Storage &signalStorage : extraSignalsStorage)
     extraSignals.emplace_back(signalStorage);
+
   if (failed(checkChannelExtra(emitError, extraSignals)))
-    return nullptr;
+    return {};
```

Comment:
Revert to `nullptr`

Reply:
I did this. Should I fix other places returning `{}` as well?

https://github.com/EPFL-LAP/dynamatic/blob/797a7d9a4303f7b63800e3f252d6ff56aa8ca3f0/lib/Dialect/Handshake/HandshakeTypes.cpp#L159-L161

https://github.com/EPFL-LAP/dynamatic/blob/797a7d9a4303f7b63800e3f252d6ff56aa8ca3f0/lib/Dialect/Handshake/HandshakeTypes.cpp#L175-L177

Reply:
Yes please, I think nullptr hints at that `Type` is really a funny pointer, and not some object we would just like to default-initialize.